src/signatures/schnorr.rs

```diff
@@ -57,8 +66,36 @@ where
 
     /// Sign a challenge with the given `secret` and private `nonce`. Returns an SchnorrSignatureError if `<K as
     /// ByteArray>::from_bytes(challenge)` returns an error.
+    ///
+    /// WARNING: The public key and nonce are NOT bound to the challenge. This method assumes that the challenge has
+    /// been constructed such that all commitments are already included in the challenge.
+    ///
+    /// Use [`sign_raw`] instead if this is what you want. (This method is a deprecated alias for `sign_raw`).
+    ///
+    /// If you want a simple API that binds the nonce and public key to the message, use [`sign_message`] instead.
+    #[deprecated(
+        since = "0.16.0",
+        note = "This method probably doesn't do what you think it does. Please use `sign_message` or `sign_raw` \
+                instead, depending on your use case. This function will be removed in v1.0.0"
+    )]
+    #[allow(clippy::needless_pass_by_value)]
     pub fn sign(secret: K, nonce: K, challenge: &[u8]) -> Result<Self, SchnorrSignatureError>
-    where K: Add<Output = K> + Mul<P, Output = P> + Mul<Output = K> {
+    where
+        K: Add<Output = K>,
+        for<'a> K: Mul<&'a K, Output = K>,
+    {
+        Self::sign_raw(&secret, nonce, challenge)
```

Comment:
Since the secret key is passed by value here, we could change the trait bound on `K` to require that it implement `Zeroize` (or modify `SecretKey` to require `Zeroize` instead), and then call `secret.zeroize()` before returning the signature. Then this library would never keep secret keys in memory.

I think a more comprehensive approach is to require that `SecretKey` implement `ZeroizeOnDrop` (if/when that marker trait becomes available to this crate), at which point you wouldn't need an explicit call to `zeroize()`, and could also provide better assurances that an implementer is handling secret keys better.

Right now it just so happens that the only implementer in this crate uses `RistrettoSecretKey`, which does zeroize on drop... but that's not a particular trait bound.

Reply:
[This issue](https://github.com/tari-project/tari-crypto/issues/147) has been opened for discussion on the blockers to adding a `ZeroizeOnDrop` bound to `SecretKey`.

Reply:
Zeroize trait bounds are a good idea, but out of scope for this PR.